mir: use `TypeId`

zerbina opened the pull request:
## Summary

Store references to types as `TypeId` instead of `PType`, making
`MirNode` a plain data-type. Besides speeding up the compiler, this
also paves the way to a data-oriented type IR for the MIR.

## Details

### Overview

* all MIR-level types use `TypeId` instead of `PType` (currently only
  `MirNode` and `Local`)
* the new `TypeEnv` type from the `mirtypes` module manages the type
  storage and mappings
* a `TypeEnv` instance is embedded in `MirEnv`
* registering a `PType` with `TypeEnv` yields a `TypeId`
* querying type information requires access to a `TypeEnv` instance

### Mapping Types

`TypeEnv` stores the `PType`-to-`TypeId` mappings in a `TypeTable`.
`TypeTable`, provided by the new `typemaps` module, is a custom
`Table` type that de-duplicates the `PType` keys by using structural
or symbolic comparison/hashing for the keys, thereby significantly
reducing the number of mappings and registered types.

At the moment, `PType` is still used as the type IR, so `TypeEnv` only
stores a `PType` for each `TypeId`.

The built-in types are mapped to statically known IDs, making some
translation and MIR construction easier.

### Constant expressions

To not require passing a `TypeEnv` to `datatables.getOrPut`, only the
type IDs are compared. This doesn't produce the same results as
`sameBackendType`, which ignores, for example, the distinct type
modifiers and aliases. As a result, `IntAlias(1)` and `int(1)` now
result in two `DataTable` entries instead of one.

Only using `TypeId` comparison/hashing does speed up `getOrPut`,
however.

### CGIR

* the CGIR still uses `PType`
* `cgirgen` turns the input `TypeId`s back into `PType`s
* since `cgirgen` creates new `PType` and assigns them to `Local`s,
  access to a *mutable* `MirEnv` is now required (in order to register
  the new `PType`s)

### VM backend

Packing constant data requires querying type information, and access to
a `TypeEnv` is thus required. A pointer to the `TypeEnv` instance is
passed along in `DataEncoder`.

### VM Evaluation Pass

Initializing the JIT environment requires a basic type environment,
which is not present when using the VM evaluation pass, as the `system`
module isn't yet processed at the time the `PEvalContext` is
initialized.

To delay setting up the JIT environment as much as necessary, the
evaluation pass is restructured:
* only a VM instance is assigned to the `ModuleGraph` at the start
* `setupGlobalCtx` creates a proper evaluation context from the
  instance
* evaluating statements, which now calls `setupGlobalCtx`, is prevented
  for declarative statements

This ensures that `initJit` is only called after the processing of the
`system` module has set up the basic type environment.

### Compiler Performance

`MirNode` being a plain data-type (no destructor) significantly speeds
up all MIR processing that creates new nodes or copies/moves them,
especially `TreeChangeset` application (~ -70%).

In addition, no copy/sink/destroy needing to be injected for `MirNode`s
also speeds up bootstrapping, since it means less work for the
destructor injection pass.

zerbina committed:
mir: reference types through `TypeId`

zerbina committed:
introduce `typemaps` module

It contains the implementation of a simple `Table` with `PType` as the
key, where the `PType` undergo basic canonicalization.

zerbina committed:
initialize the `MirEnv` everywhere

`MirEnv` is now a type that requires explicit initialization, meaning
that all types it's embedded within do too.

zerbina committed:
add the `mirtypes` module

In the future, it'll contain the type definitions for the type IR, but
for the moment, only the type environment type `TypeEnv` is and its
setup is located in it.

A instance of `TypeEnv` is directly embedded into `MirEnv`. Since
`TypeEnv` is a type requiring initialization, `MirEnv` too requires
explicit initialization, which is provided via `initMirEnv`.

zerbina committed:
utils: update MIR pretty-printer

* without access to a `MirEnv`, only the type IDs are renderer
* for fields, without access to a `MirEnv`, "Field" + field position
  is used

zerbina committed:
mirgen: use `TypeId`

* the `mirconstr` API now only accepts `TypeId`
* `mirgen` registers the input `PType` with the environment through the
  `typeToMir` procedure
* `typeOrVoid` is obsolete and thus removed

zerbina committed:
mirgen: remove  workaround

An untyped `nkNilLit` reaching `mirgen` has long been fixed, the
workaround is obsolete.

zerbina committed:
adjust the MIR passes

* a `TypeEnv` has to be available where `PType` needs to be inspected
* where possible, the type ID constants are used

zerbina committed:
backends: adjust MIR production

The `mirconstr` API only accepts `TypeId` now, meaning that the `PType`s
have to be translated first (by adding them to the environment). Where
possible, the type IDs for built-in types are used directly.

zerbina committed:
mirgen: don't register template's type

The `tyProc` type of a template possibly contains generic types, which
`typemaps` doesn't - and shouldn't need to - support. Therefore, the
AST value for the template symbol gets `VoidType` assigned. Ultimately,
the type of the AST value doesn't matter, as `vmgen` special-cases it
any way.

zerbina committed:
adjust to `TypeId` usage in `ConstrTree`

zerbina committed:
datatables: adjust to `TypeId` usage

Instead of comparing the `PType`s, the `TypeId`s are compared. This
isn't exactly equivalent, as `sameBackendType` ignores type modifiers
and wrapper types, whereas a `TypeId` comparison does not.

Using `sameBackendType` would require passing along the `TypeEnv`
however, so the `TypeId` comparison is preferred.

zerbina committed:
cgirgen: translate type IDs back to `PType`

The CGIR still uses `PType` (and will continue to do so), so the type
IDs have to be translated back.

Since `cgirgen` creates new types (var and lent), it now requires
access to a mutable `MirEnv`.

zerbina committed:
packed_env: adjust to `TypeId` usage in `ConstrTree`

Unfortunately, packing the constant data requires querying information
about types, which requires access to a `TypeEnv`. A pointer to the
instance is passed along in `DataEncoder`.

zerbina committed:
cgen, jsgen: adjust to `Local` using `TypeId`

zerbina committed:
vmgen: adjust to `Local` using `TypeId`

The `usesRegister` procedure needs access to a `MirEnv`, for fetching
the `PType` for a `TypeId`.

zerbina committed:
prevent int/float literal types from being the canonical one

The first `PType` added to the environment for type X is treated as the
canonical one, which is a problem if said `PType` is an integer/float
literal type (all integer types would then show as integer/float
literal types).

To work around this, the `system.int` type is registered right away,
establishing it as the canonical `tyInt` type.

zerbina committed:
tests: adjust `tdatatables.nim`

zerbina committed:
delay JIT creation in scripting context

Initializing the JIT environment requires a basic type environment, but
this is not present when using the VM evaluation pass, as the `system`
module hasn't been processed at the time the `PEvalContext` is
initialized.

To delay setting up the JIT environment as much as possible, the
evaluation pass is restructured:
* only a VM instance is assigned to the `ModuleGraph` at the start
* `setupGlobalCtx` creates a proper evaluation context from the
  instance
* evaluating statements, which now calls `setupGlobalCtx`, is prevented
  for declarative statements

This ensures that `initJit` is only called after the processing of the
`system` module has progressed far enough.

saem commented on compiler/backend/compat.nim:
```diff
@@ -146,20 +147,22 @@ proc translate*(t: MirTree, env: MirEnv): CgNode =
       ## Convenience template for setting up the tree node and iterating the
       ## input node's child nodes.
-      let res {.inject.} = newExpr(k, unknownLineInfo, n.typ)
+      let res {.inject.} = newExpr(k, unknownLineInfo, typ)
```
No change required, just a bit weird that `typ` is used before it's definition below.

zerbina replied:
It is. I do prefer using either early binding or dedicated parameters for templates, and there's nothing speaking against using early binding here -- I'll change it.

saem commented on compiler/mir/mirgen.nim:
```diff
@@ -1150,42 +1148,33 @@ proc genSetConstr(c: var TCtx, n: PNode) =
     # need
 
-    c.subTree (if isConsume: mnkConsume else: mnkArg): # the environment
-      if n[1].kind == nkNilLit:
-        # it can happen that a ``nkNilLit`` has no type (i.e. its typ is nil) -
-        # we ensure that the nil literal has the correct type
-        # TODO: prevent a ``nkNilLit`` with no type information from being
-        #       created instead
-        c.add MirNode(kind: mnkNilLit,
-                      typ: c.graph.getSysType(n[1].info, tyNil))
```
Sorry, might be a silly question, but when did the change to ensure `nkNilLit` had a `typ` field happen? IIUC, this should be sem fix but I can't recall it. Maybe not quite which specific commit or anything, but more was this a series of fixes over time and became vestigial, or was there something specific?

zerbina replied:
No, not a silly question. This removal was originally intended to be a separate PR, but it felt to minuscule to warrant the PR overhead.

This wasn't an issue with sem, rather with `transf` (as only there are `nkClosure` nodes created), but I don't remember which change exactly fixed this bug. If I remember correctly, the problem was related to the JavaScript backend being selected, so *believe* it was #586 that fixed it.

saem commented on compiler/mir/typemaps.nim:
```diff
@@ -0,0 +1,187 @@
+  t.inner[Type key]
+
+proc `[]=`*[T](t: var TypeTable[T], key: PType, val: sink T) {.inline.} =
+  t.inner[Type key] = val
```
```suggestion
  ## Assign a `val` for the `key`.
  t.inner[Type key] = val
```

saem commented on compiler/mir/typemaps.nim:
```diff
@@ -0,0 +1,187 @@
+  if n != nil: n.id
+  else:        NodeId -1
+
+proc cmp(a, b: PType): bool {.noSideEffect.}
```
Why not `func`?

zerbina replied:
No specific reason. The compiler complained (rightfully so) about that calling the forward-declared `cmp` had side effects, so I mindlessly added the pragma. I'll change it to use `func`.

saem reviewed (approved):
Looks good.

All comments are minor/non-blocking.

zerbina committed:
typemaps: add doc comment to `[]=`

Co-authored-by: Saem Ghani <saemghani+github@gmail.com>

zerbina committed:
typemaps: make all private procedures `func`s

zerbina committed:
refactor(compat): move variable definitions around

This makes the code a bit easier to read, since `typ` is now defined
before it's first appearance.

saem commented:
/merge

zerbina commented:
Thank you for the review, @saem.
